sklearn/cluster/_kmeans.py

```diff
@@ -1572,14 +1593,16 @@ def fit(self, X, y=None, sample_weight=None):
             Fitted estimator.
         """
         self._validate_params()
-        engine = self._get_engine(X, y, sample_weight, reset=True)
+        engine, validated_inputs = self._get_engine(X, y, sample_weight, reset=True)
 
+        X = validated_inputs["X"]
+        y = validated_inputs["y"]
+        sample_weight = validated_inputs["sample_weight"]
```

Comment:
Too magic?

```suggestion
        # Re-assign the X, y and sample_weight local variables
        locals().update(validated_inputs)
```

no opinion, feel free to mark as resolved if you don't like it but this would avoid quite some verbose duplication.

Reply:
What if `validated_inputs` contains more than those 3 items? Do we want to store references to those in/via `locals()`?

Reply:
I think it's the responsibility of the plugin to not mess around with the estimator's internal.